### PR TITLE
Ports - Elle & Sopheary

### PR DESCRIPTION
**Pull request by dev-elle-up**
# Adagrams
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the components that make up a method? | The components of a method are the "def" keyword, method signature (including the method name and any arguments), code block, and end keyword.
What are the advantages of using git when collaboratively working on one code base? | Git enables you to track changes and go back to previous states of the code if something breaks in the future. If the code is stored on a remote repository, multiple team members can share the project files by cloning the project to their local machines. They can each work on different parts of the code, or work on it at different times, and then merge changes efficiently. 
What kind of relationship did you and your pair have with the unit tests? | We ran the unit tests frequently after building each method. We used the resulting failures and errors to fix our code. We did this one error at a time, running the spec file after each change.
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | Yes! It uses .map and .min_by. ".map" was helpful to quickly and efficiently create a new array with all the words having the max score. We could then iterate through that array to find the winning word. ".min_by" let us find the highest scoring word with the fewest letters in case of a tie. 
What was one method you and your pair used to debug code? | Throughout the "uses_available_letters" method, we entered several puts statements which showed us what was happening to the arrays as the code ran. This was essential for our success.
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | 1) Switching every 20 minutes was frustrating and anxiety inducing for both of us because it broke our rhythm just as we would get into a groove. When we adjusted our period of time to switch between driver and navigator to be longer (30 minutes with an optional extra 5 minutes to finish a thought) worked much better. 2) We liked that we both learned things from each other.

### CODE REVIEW

**Comment by tildeee** [comment]
# Adagrams
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | x
Both teammates contributed to the codebase | x
Small commits with meaningful commit messages | x, in general, we would like have commit messages that described the changes, rather than "wave 2" etc.
**Code Requirements** |
`draw_letters` method | x
Uses appropriate data structure to store the letter distribution | x
All tests for `draw_letters` pass | x
`uses_available_letters?` method | x
All tests for `uses_available_letters?` pass | x
`score_word` method | x
Uses appropriate data structure to store the letter scores | x
All tests for `score_word` pass | x
`highest_score_from` method | x
Appropriately handles edge cases for tie-breaking logic | x
All tests for `highest_score_from` pass | x
**Overall** |

Hi Elle and Sopheary! Good work on this project!

Your submitted code is good code style, and proves to me that the iteration and use of `Enumerable` methods is solid. Great job with getting all of the implementation done well and correctly

I have a few comments about some things like global variables and a helpful method to use in the future to make your code slimmer

That being said, great work on this project overall!

File: lib/adagrams.rb [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+  X: 1,
+  Y: 2,
+  Z: 1,
+}
+$letter_pool = []
+letter_quantities.each do |letter, quantity|
+  quantity.times do
+    $letter_pool << letter.to_s
+  end
+end
```
Comment (tildeee) [inline lib/adagrams.rb:34]: You all populated a variable representing the letter pool as a global variable! In some ways, this makes sense, but in general, we'll avoid using global variables. Our hope was top put the above few lines within the `draw_letters` method, since that's the only time you'll use this variable.

File: lib/adagrams.rb [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+$letter_scores = {
+  1 => ["A", "E", "I", "O", "U", "L", "N", "R", "S", "T"],
+  2 => ["D", "G"],
+  3 => ["B", "C", "M", "P"],
+  4 => ["F", "H", "V", "W", "Y"],
+  5 => ["K"],
+  8 => ["J", "X"],
+  10 => ["Q", "Z"],
+}
```
Comment (tildeee) [inline lib/adagrams.rb:44]: Similar to my previous comment: we'll try to avoid using global variables in the future, and the only method that uses this variable `$letter_scores` is in the `score_word` method, so we could move this variable there.

File: lib/adagrams.rb [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+  return letters_in_hand
+end
+
+# Wave 2
+def uses_available_letters?(input, letters_in_hand)
+  input = input.upcase.split("")
+  # Make a copy of the input array and keep the original 
+  # array unchanged. The original input needs to remain in tact so
+  # that Ruby can continue to iterate through it the 
+  # uses_available_letters methods.
```
Comment (tildeee) [inline lib/adagrams.rb:58]: This is a great use of a comment: to add detail about something that feels a little weird and needs explanation!

File: lib/adagrams.rb [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    
+    if letters_in_hand.include?(letter)
+      hand_index = 0
+      letters_in_hand.each_with_index do |letter_in_hand, index|
+        if letter == letter_in_hand
+          hand_index = index
+          break
+        end
+        hand_index += 1
+      end
```
Comment (tildeee) [inline lib/adagrams.rb:72]: Nice use of `each_with_index`! It seems to me that the purpose of this loop is to find the index of the letter you're looking for. There is a method that works on strings for this: `letters_in_hand.index(letter)`... which will give you the first index of a matching element (`letter`) in the array (`letters_in_hand`). It's assuming to find the _first_ index of a matching element, but you can safely make that assumption since you're iterating through `input` in order :) (Actually, it kind of feels like you _implemented_ the `.index` method, so great job on that :D)

File: lib/adagrams.rb [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+      letters_in_hand.delete_at(hand_index)
```
Comment (tildeee) [inline lib/adagrams.rb:74]: In conjunction with what I mentioned above, you could replace this line with: `letters_in_hand.delete_at(letters_in_hand.index(letter))`

File: lib/adagrams.rb [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+      i = 0
+      input_copy.each_with_index do |input_copy_letter, input_copy_index|
+        if input_copy_letter == letter
+          i = input_copy_index
+          break
+        end
+      end
+      input_copy.delete_at(i)
```
Comment (tildeee) [inline lib/adagrams.rb:83]: This loop ends up solving a similar problem to above, and therefore, has a similar solution. With that, if you wanted to use the `.index` method, you could replace this line with: `input_copy.delete_at(input_copy.index(letter))`

File: lib/adagrams.rb [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        end
+      end
+    end
+  else
+    winning_word[:word] = winning_words[0]
+  end
+
+  winning_word[:score] = max_score.to_i
+  return winning_word
+end
```
Comment (tildeee) [inline lib/adagrams.rb:151]: This method ended up being pretty long -- if methods get this long, feel free to break sections up into smaller helper methods
